Add Chat Template Support to vLLM

Tostino opened the pull request:
This pull request introduces the chat template feature to vLLM, utilizing the template stored in the tokenizer, enhancing its compatibility with the OpenAI Chat API.
https://huggingface.co/blog/chat-templates

Key Changes:

1. Chat Template Integration: Implemented the ability for the vLLM server to utilize a predefined jinja chat template stored in the tokenizer for supported models.
2. Template Override: Added functionality that lets users provide their custom chat templates using the --chat-template argument.
3. Documentation Update: The Quickstart guide and related documentation have been updated to instruct users on how to use this new feature.

Benefits:

1. Replaces FastChat with a more standardized way of handling templates.
2. New models or templates don't need special support added to another library, they just need to implement the template in the tokenizer.
3. Fixes issues with formatting being inconsistently applied.

Would appreciate reviews and feedback on this new feature integration.

Tostino committed:
Add support for HF chat templates to OpenAI chat completions API. Add better documentation as well.

Tostino commented:
I've not touched any of the code that is failing the pylint error.

Tostino committed:
Change variable name to fix pylint error.

flexchar commented:
This would be so good. I used Mistral with vLLM and Chat Completion and it was using horribly wrong template from fastchat. It must have been ChatML, it was something with hashtags and roles.

AguirreNicolas commented:
I found this functionality really necesary to all. Please consider to support it. 

So far I tested with 

* TheBloke/Llama-2-13B-chat-AWQ
* TheBloke/Mistral-7B-OpenOrca-AWQ
* TheBloke/zephyr-7B-beta-AWQ

## TheBloke/Llama-2-13B-chat-AWQ
The chat_template wasn't in the `template_config.json`, then I use `--chat-template` flag pointing to the official `chat_template` from the meta's `template_config.json`. It worked properly.

## TheBloke/Mistral-7B-OpenOrca-AWQ
The `chat_template` is already in `template_config.json`, so the calls were smooth and ok. 

## TheBloke/zephyr-7B-beta-AWQ
The `chat_template` is already in `template_config.json`, **BUT** when calling to `http://localhost:8000/v1/chat/completions`, there is a `<|assistant|>\n` added  at the begginin of the `content` of the. 
So I will test it further. Maybe the default `chat_template` is no properly defined, or maybe is some eror in the code.

Hope you find usefull the feedback.

timothylimyl commented:
> I found this functionality really necesary to all. Please consider to support it.
> 
> So far I tested with
> 
>     * TheBloke/Llama-2-13B-chat-AWQ
> 
>     * TheBloke/Mistral-7B-OpenOrca-AWQ
> 
>     * TheBloke/zephyr-7B-beta-AWQ
> 
> 
> ## TheBloke/Llama-2-13B-chat-AWQ
> 
> The chat_template wasn't in the `template_config.json`, then I use `--chat-template` flag pointing to the official `chat_template` from the meta's `template_config.json`. It worked properly.
> ## TheBloke/Mistral-7B-OpenOrca-AWQ
> 
> The `chat_template` is already in `template_config.json`, so the calls were smooth and ok.
> ## TheBloke/zephyr-7B-beta-AWQ
> 
> The `chat_template` is already in `template_config.json`, **BUT** when calling to `http://localhost:8000/v1/chat/completions`, there is a `<|assistant|>\n` added at the begginin of the `content` of the. So I will test it further. Maybe the default `chat_template` is no properly defined, or maybe is some eror in the code.
> 
> Hope you find usefull the feedback.

did you manage to add your own custom system message to Zephyr?

I raised an issue here: [https://github.com/vllm-project/vllm/issues/1530](Here)

Tostino commented:
@WoosukKwon is there a chance to get a review on this?

adamlin120 commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -70,50 +62,17 @@ async def check_model(request) -> Optional[JSONResponse]:
+    if chat_template is not None:
+        return tokenizer.apply_chat_template(conversation=request.messages,
+                                             chat_template=chat_template,
+                                             tokenize=False)
```
@Tostino  should we add add_generation_prompt=True for generation prompt?

Tostino replied:
Uh, yeah I am pretty sure we should. Let me do a little research tomorrow on templates that properly support it. The template needs to actually use that variable somewhere in it for it to take any effect. I had a long travel day today, will get back to you tomorrow some time.

aarnphm replied:
afaik `add_generation_prompt` is currently only needed for mistral (or at least I tested with mistral). For llama you don't need add_generation_prompt

Probably should allow user to control this, maybe through another cli flag `--add-generation-prompt/--no-add-generation-prompt` (default to False)?

What do you think?

Tostino replied:
Okay, I believe I have resolved this now with the latest commit. This is now part of the request. I agree that we should allow users to set default values for those parameters through CLI flags, but think we should do that in it's own PR for all of the parameters that make sense.

casper-hansen commented:
One thing that I don’t understand is why chat templates were integrated to begin with. It introduces a lot of untransparent bugs, especially if you use a non-standard format. 

How can I use vLLM without any chat template applied where the chat template is applied directly in the prompt that I pass to the engine?

AguirreNicolas commented:
> One thing that I don’t understand is why chat templates were integrated to begin with. It introduces a lot of untransparent bugs, especially if you use a non-standard format. 
> 
> How can I use vLLM without any chat template applied where the chat template is applied directly in the prompt that I pass to the engine?

This PR only apply to `vllm/entrypoints/openai/api_server.py`
You should use  `vllm/entrypoints/api_server.py` instead.

Tostino committed:
Merge branch 'main' into chat_templates

# Conflicts:
#	vllm/entrypoints/openai/api_server.py

dongxiaolong commented:
Hello  @WoosukKwon  and  @zhuohan123 

We are facing an urgent issue that requires immediate attention and resolution. I hope you can review and merge the associated PR to address this problem. The issue at hand is causing erroneous and extraneous data to be included in the model's input, such as:

_"### Human: Got any creative ideas for a 10 year old’s birthday?
### Assistant: Of course! Here are some creative ideas for a 10-year-old's birthday party:
1. Treasure Hunt: Organize a treasure hunt in your backyard or nearby park. Create clues and riddles for the kids to solve, leading them to hidden treasures and surprises.
2. Science Party: Plan a science-themed party where kids can engage in fun and interactive experiments. You can set up different stations with activities like making slime, erupting volcanoes, or creating simple chemical reactions.
3. Outdoor Movie Night: Set up a backyard movie night with a projector and a large screen or white sheet. Create a cozy seating area with blankets and pillows, and serve popcorn and snacks while the kids enjoy a favorite movie under the stars.
4. DIY Crafts Party: Arrange a craft party where kids can unleash their creativity. Provide a variety of craft supplies like beads, paints, and fabrics, and let them create their own unique masterpieces to take home as party favors.
5. Sports Olympics: Host a mini Olympics event with various sports and games. Set up different stations for activities like sack races, relay races, basketball shooting, and obstacle courses. Give out medals or certificates to the participants.
6. Cooking Party: Have a cooking-themed party where the kids can prepare their own mini pizzas, cupcakes, or cookies. Provide toppings, frosting, and decorating supplies, and let them get hands-on in the kitchen.
7. Superhero Training Camp: Create a superhero-themed party where the kids can engage in fun training activities. Set up an obstacle course, have them design their own superhero capes or masks, and organize superhero-themed games and challenges.
8. Outdoor Adventure: Plan an outdoor adventure party at a local park or nature reserve. Arrange activities like hiking, nature scavenger hunts, or a picnic with games. Encourage exploration and appreciation for the outdoors.
Remember to tailor the activities to the birthday child's interests and preferences. Have a great celebration!"_

This unnecessary information introduces significant instability to the user experience. It's critical that we address this to maintain the integrity and reliability of our model.

Thank you for your prompt action on this matter.

aarnphm commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -31,20 +31,13 @@
 from vllm.transformers_utils.tokenizer import get_tokenizer
 from vllm.utils import random_uuid
 
-try:
-    import fastchat
-    from fastchat.conversation import Conversation, SeparatorStyle
-    from fastchat.model.model_adapter import get_conversation_template
-    _fastchat_available = True
-except ImportError:
-    _fastchat_available = False
```
Since this PR drops support for fastchat can you also take a quick search to see if there is any additional fastchat import anywhere else to remove it?

Tostino replied:
Yes, there are no other references that I found.

aarnphm reviewed (commented):
LGTM, with one comment. Will need @WoosukKwon to signoff

dongxiaolong commented:
Hi@Tostino.Have you noticed that the current template causes some models to occasionally return tags like <|im_start|>?

Tostino commented:
> LGTM, with one comment. Will need @WoosukKwon to signoff

I don't know if it's good yet. I looked at things a few nights ago, but am on vacation right now and couldn't spend enough time to verify how it works with multiple models (slow net speed where I am right now).

I don't think it should be passed as a CLI argument. Maybe as part of the API request, but I need some more time to actually test things out.

Reason being, you want to be able to have the endpoint complete responses that are partially started, and I wanted to make sure that is still possible with various templates already in use.

The other thing is, not all templates out there actually follow the spec. I'd say we should make those models change their template to work properly rather than working around the bugs in the template. I was planning on documenting the primary templates used today and the models using them to see what is valid, and what needs more work.

aarnphm commented:
> I don't think it should be passed as a CLI argument. Maybe as part of the API request, but I need some more time to actually test things out.

Chat template should be a before-inference options right? If you are running the models that only works with one chat template then I don't really see the point of changing prompt template on inference

Tostino commented:
> > I don't think it should be passed as a CLI argument. Maybe as part of the API request, but I need some more time to actually test things out.
> 
> Chat template should be a before-inference options right? If you are running the models that only works with one chat template then I don't really see the point of changing prompt template on inference

I'm not saying to change the template at inference time, just that it may make sense for some inference requests to need to pass in add_generation_prompt=True and others will want to pass in False if they are wanting a completion of an existing prompt template if the model is trained to do that. I wanted to test that prior to locking us into a design that is wrong.

aarnphm commented:
Yes that makes sense. But I still see the value of setting it at CLI as well, in additional to the ability to change at inference time

Tostino committed:
Merge branch 'vllm-project:main' into chat_templates

Tostino committed:
Added 'add_generation_prompt' Request Parameter (Default: True): This parameter controls whether the prompt ends with tokens indicating the start of an assistant message. When set to False (and the template/model supports it) the model should complete the last response in the list. By default, it maintains compatibility with OpenAI's API behavior.

Fixed Role Determination in Responses: Resolved an issue where the role in responses defaulted to "assistant" regardless of context. This fix ensures the response role aligns with the intended conversational participant, enhancing the API's versatility in various chat scenarios.

Introduced 'return_full_response' Request Parameter (Default: False): This parameter, when set to True, negates the need for client-side response merging. It simplifies client integration by providing complete responses even when the client had started the response for the model.

Tostino committed:
Merge branch 'vllm-project:main' into chat_templates

aarnphm commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -70,50 +62,21 @@ async def check_model(request) -> Optional[JSONResponse]:
+            chat_template=chat_template,
+            tokenize=False,
+            add_generation_prompt=request.add_generation_prompt)
+    elif tokenizer.chat_template is not None:
```
```suggestion
    elif hasattr(tokenizer, 'chat_template') and tokenizer.chat_template is not None:
```

We can probably clean this up and just apply the following

```python
if chat_template is not None:
  tokenizer.chat_template = chat_template

tokenizer.apply_chat_template(...)
```

Tostino replied:
Sure, I had actually done that originally, but then noticed the value in the constructor...will go back to that implementation.

aarnphm reviewed (commented):
Also I don't think return_full_response makes sense here. The purpose of this file is to makes OpenAI compatible API. I don't think people who uses OpenAI will have the expectation to return the generated text with prompt.

My philosophy on this is since it is a compatible server, we shouldn't deviate from the ground truth too much

Tostino commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -629,6 +631,13 @@ async def fake_stream_generator() -> AsyncGenerator[str, None]:
                               tokenizer_mode=engine_args.tokenizer_mode,
                               trust_remote_code=engine_args.trust_remote_code)
 
+    if chat_template or tokenizer.chat_template:
```
Huh yeah, thanks for pointing that out: https://github.com/huggingface/transformers/blob/main/src/transformers/tokenization_utils_base.py#L1795

These things change so quick!

Tostino replied:
Okay, have changed this to still give a warning if undefined, and print out the template in use as in info log.
```
    if chat_template is not None:
        tokenizer.chat_template = chat_template

    if tokenizer.chat_template or tokenizer.default_chat_template:
        logger.info(
            f"Chat template:\n{tokenizer.chat_template or tokenizer.default_chat_template}")
    else:
        logger.warning(
            "No chat template loaded, the chat endpoint will be disabled.")
```

aarnphm replied:
Wdym by `will be disabled` here? it won't use any chat template right?

Now thinking more about this, I feel like the decision of formatting chat templates or not should leave it to the users.

Especially for people who just pass the tokens via `prompt_token_ids` instead of using `prompt` parameters.

Tostino replied:
In essence, the chat completions API needs a template for formatting input. You aren't converting from a list of messages (role/content pairs)  to a formatted piece of text without *something* driving what to do. Without a template, it will error out. However, as long as you're not using an early version of Transformers that supports chat templates but not the default template, you should be fine...

Regarding the `prompt_token_ids`, they are not applicable in the chat completions API since it requires a message list consisting of strings only. Integers not allowed.

In my view, choosing to use the chat API itself is an opt-in decision. The chat API inherently needs a template to function correctly, involving some necessary internal/external configurations for how that template should work on the server side.

If users prefer client-side formatting, the regular completions API is more suitable. It allows the submission of pre-formatted `prompt` or `prompt_token_ids`, offering greater flexibility in formatting.

Tostino commented:
> Also I don't think return_full_response makes sense here. The purpose of this file is to makes OpenAI compatible API. I don't think people who uses OpenAI will have the expectation to return the generated text with prompt.
> 
> My philosophy on this is since it is a compatible server, we shouldn't deviate from the ground truth too much

The OpenAI chat completion API doesn't support completions at all, so you can't start a response for the assistant (or the user) and have it finish it. You can do that with plenty of local models, and that is the point of the `add_generation_prompt` parameter to begin with. When you add that support, it makes sense to make this easier on the clients to support both the existing real OpenAI endpoint which doesn't support this functionality (thus wouldn't benefit from this feature) and those who have migrated to local models which support more features.

How would you suggest packing up parameters which are only needed for some models/templates which support them?

Tostino commented:
In my testing, I also found inconsistencies in how the streaming api was implemented with the OpenAI implementation. Parts of the json that should have been excluded if there was no data were always being included. I have a fix for that in the works.

Tostino committed:
Renamed `return_full_response` to `echo`, and made it also work with streaming responses after testing how it worked with the regular OpenAI completion API.

Fixed inconsistencies with official OpenAI API, and what we were returning for streaming responses for the chat completion API.

Added error handling so if there is an issue with applying the template, it is reported to the user through an API error, and logged.

Tostino commented:
Okay, changes I was working on today were just pushed. Error handling in place. If you have a template that errors, it will report the error properly. Fixed issues with the streaming API where it didn't conform to the spec. Renamed `return_full_response` to `echo`, and implemented it for streaming (as I should have originally)

aarnphm commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -619,6 +620,28 @@ async def fake_stream_generator() -> AsyncGenerator[str, None]:
+            try:
+                # If opening a file fails, set chat template to be args to
+                # ensure we decode so our escape are interpreted correctly
+                chat_template = codecs.decode(args.chat_template,
+                                              "unicode_escape")
```
shouldn't we use jinja as well? seems like HF uses jinja template right?

Tostino replied:
This is still jinja, it just properly parses the escaped characters (`\n` and the like).

It allows users to pass in the template like so:
```
python3 -m vllm.entrypoints.openai.api_server --model Tostino/Inkbot-13B-8k-0.2 --tensor-parallel-size 2 --chat-template "<#meta#>\n- Date: {{ (messages|selectattr('role', 'equalto', 'meta-current_date')|list|last).content|trim if (messages|selectattr('role', 'equalto', 'meta-current_date')|list) else '' }}\n- Task: {{ (messages|selectattr('role', 'equalto', 'meta-task_name')|list|last).content|trim if (messages|selectattr('role', 'equalto', 'meta-task_name')|list) else '' }}\n<#system#>\n{{ (messages|selectattr('role', 'equalto', 'system')|list|last).content|trim if (messages|selectattr('role', 'equalto', 'system')|list) else '' }}\n<#chat#>\n{% for message in messages %}\n{% if message['role'] == 'user' %}\n<#user#>\n{{ message['content']|trim -}}\n{% if not loop.last %}\n\n{% endif %}\n{% elif message['role'] == 'assistant' %}\n<#bot#>\n{{ message['content']|trim -}}\n{% if not loop.last %}\n\n{% endif %}\n{% elif message['role'] == 'user_context' %}\n<#user_context#>\n{{ message['content']|trim -}}\n{% if not loop.last %}\n\n{% endif %}\n{% endif %}\n{% endfor %}\n{% if add_generation_prompt and messages[-1]['role'] != 'assistant' %}\n<#bot#>\n{% endif %}"
```

aarnphm replied:
For chat template I think it is good to recommend for users to write it into a file, rather than maintaining on the CLI. No need to do two things when there is one good way of doing it 😄

Tostino replied:
Are you sure we should lock in with file-on-disk with all the more recent "serverless" type deployment methods, container images, etc?  It allows the model's template to be copy/pasted from the tokenizer_config.json into the CLI argument, and modified as necessary without any extra steps. It's easy enough to remove, just seemed like a nice QOL improvement.

aarnphm replied:
You are right. Actually dumping it into a envvar wouldn't be a bad idea overall.

Tostino replied:
Agreed there. It alright to do that in another PR? There are likely other envvars we would want to add support for at the same time.

aarnphm replied:
I think this is pretty inline with this feature right? cc @WoosukKwon for input in this here.

Tostino replied:
Is this conversation alright to mark "resolved"?

simon-mo replied:
I think file is a good first pass here.

simon-mo commented:
@Tostino thank you for this PR. Do you know whether there are default ChatTemplate for all the model architecture vLLM supports today?

aarnphm commented:
> @Tostino thank you for this PR. Do you know whether there are default ChatTemplate for all the model architecture vLLM supports today?

afaik only mistral, llama, baichuan, stablelm, phi has a default chat templates

simon-mo commented:
In that case, maybe we can use fastchat as a fallback?

aarnphm commented:
> In that case, maybe we can use fastchat as a fallback?

I think by default if any models doesn't have a default chat templates, transformers>4.35.0 will fallback to a default one (IIRC it is the chatml template)

Tostino commented:
@simon-mo here are all the models that have their own default chat templates through HF: https://github.com/search?q=repo%3Ahuggingface/transformers%20default_chat_template&type=code

Fluder-Paradyne commented on docs/source/getting_started/quickstart.rst:
```diff
@@ -129,3 +138,45 @@ Since this server is compatible with OpenAI API, you can use it as a drop-in rep
+    $         "model": "facebook/opt-125m",
+    $         "messages": [
+    $             {"role": "system", "content": "You are a helpful assistant."},
+    $             {"role": "user", "content": "Who won the world series in 2020?"},
```
ran the curl, have to remove `,` to run

Tostino replied:
Will get that fixed, thanks for pointing it out.

Tostino commented:
Looks like there are now conflicts to resolve after recent merges/release. I'll get that done as soon as I can...

simon-mo commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -619,6 +620,28 @@ async def fake_stream_generator() -> AsyncGenerator[str, None]:
+                # ensure we decode so our escape are interpreted correctly
+                chat_template = codecs.decode(args.chat_template,
+                                              "unicode_escape")
+            except json.JSONDecodeError:
+                logger.error("Unable to set template.")
```
should we hard fail here? in this case users passed in a chat template but we really can't set it.

Tostino replied:
The other endpoints that don't rely on the template can still be available, and the user will get an error returned in their json response if they try and hit the chat endpoint and it fails.

simon-mo commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -619,6 +620,28 @@ async def fake_stream_generator() -> AsyncGenerator[str, None]:
     else:
         served_model = args.model
 
+    if args.chat_template is not None:
```
can you move the try except into a function? thank you making it robust!

Tostino replied:
Yeah, will do.

Tostino replied:
This is done in the new PR

simon-mo commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -629,6 +652,16 @@ async def fake_stream_generator() -> AsyncGenerator[str, None]:
                               tokenizer_mode=engine_args.tokenizer_mode,
                               trust_remote_code=engine_args.trust_remote_code)
 
+    if chat_template is not None:
+        tokenizer.chat_template = chat_template
+
+    tmp_template = tokenizer.chat_template or tokenizer.default_chat_template
+    if tmp_template:
+        logger.info(f"Chat template:\n{tmp_template}")
+    else:
+        logger.warning(
+            "No chat template loaded, the chat endpoint will be not work.")
+
```
```suggestion
    if chat_template is not None:
        tokenizer.chat_template = chat_template

    final_template = tokenizer.chat_template or tokenizer.default_chat_template
    if final_template:
        logger.info(f"Using chat template:\n{tmp_template}")
    else:
        raise ValueError("No chat template loaded, the chat endpoint will be not work. There is no default template from the tokenizer. Please explicitly provide one through `--chat-template`."

```

I would prefer hard failing here to make sure there's a consistent experience as before when fastchat is not found. However this does mean if users want to use completion API only they won't be proceed forward. Perhaps we can add another flag to enable completion API with chat template?

Tostino replied:
If you prefer a hard-fail, that's fine with me. I was just going with the approach of returning an error to the user, logging the error, and allowing the other endpoints to work because it seemed like a better user experience. Will go with a hard-fail instead.

simon-mo replied:
Oh maybe hard fail and `--completion-api-only` flag. What do you think? My thinking is that fast failing on the server side is better than users getting errors at runtime for Chat endpoint.

Tostino replied:
I'll do an "endpoints" list as input, with an empty list meaning "all". I figure more endpoints will be added eventually, considering the OpenAI API expansion with agents and all that.

Any opposition?

simon-mo replied:
Sounds good!

aarnphm replied:
btw https://github.com/huggingface/transformers/blob/38e2633f80a4924bf613b0240622492beee4cfcc/src/transformers/tokenization_utils_base.py#L1731

Tostino replied:
@simon-mo So hard-fail here is actually harder than I was hoping. We don't know that the template is broken until we go to apply it at runtime due to a users request. So the template can be totally broken Jinja, and we won't know it until the user is making requests against it.

Do you want to check the template at startup validity using the jinja library directly?

With the changes to transformers to add a default template to the tokenization_utils_base, there really isn't a case I could get it to crash due to a missing tokenizer. I am just going to remove the "else" entirely at this point.

I suppose i'll leave the code in place for:
```
parser.add_argument("--disable-endpoints",
                        type=json.loads,
                        default=[],
                        help="List of disabled endpoints")
```

simon-mo commented on vllm/entrypoints/openai/protocol.py:
```diff
@@ -73,6 +73,8 @@ class ChatCompletionRequest(BaseModel):
     stop_token_ids: Optional[List[int]] = Field(default_factory=list)
     skip_special_tokens: Optional[bool] = True
     spaces_between_special_tokens: Optional[bool] = True
+    add_generation_prompt: Optional[bool] = True
```
Can you explain this parameter? It is not part of the official API request protocol.

Tostino replied:
Please see here: https://huggingface.co/docs/transformers/chat_templating#what-are-generation-prompts

It's part of the HF chat template definition. We had a discussion about it already up here: https://github.com/vllm-project/vllm/pull/1493#issuecomment-1806523808

simon-mo reviewed (commented):
I believe there's conflict with the recently merged `usage` PR. Thank you again for doing this and I think this definitely helps simplify things.

zhuohan123 commented:
Do you think it is possible/necessary to keep fastchat as an optional dependency for now? Sometimes people may wanna run  chatbots like vicuna, which does not have a template on huggingface.

simon-mo commented:
@zhuohan123 I think alternatively we can provide the Vicuna template in the examples directories so people can download them easily?

Tostino commented:
> @zhuohan123 I think alternatively we can provide the Vicuna template in the examples directories so people can download them easily? 

Completely agree here. I really don't think keeping FastChat as a dependency if we can help it.

If anyone wants to add any like Vicuna to this thread as jinja templates, I'll add them to the examples folder. Ideally, you would open a PR on the HF repo to add the template to the models. That is the long term solve for this.

Tostino commented:
Heh, well...that didn't work. 
I had to just do my rebase work in another, freshly created, branch (chat_templates_redux). It simply was way too confusing on the existing branch to resolve the conflicts multiple times for different commits that had conflicts.
I was going to try renaming the old chat_templates branch -> chat_templates_old and then chat_templates_redux -> chat_templates...it ended up closing this PR.

Give me a few... trying to get it fixed if I can, if not I will just open a new PR =(.

Tostino commented:
Well crap, from what I read I can't fix at this point, and need to open a new PR: https://github.com/vllm-project/vllm/pull/1756

@simon-mo and @aarnphm  ^^^

aarnphm commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -70,50 +63,13 @@ async def check_model(request) -> Optional[JSONResponse]:
 
 
 async def get_gen_prompt(request) -> str:
-    if not _fastchat_available:
```
cc @Tostino this is the previous behaviour.

Stealthwriter commented:
I've been following this for a while, I'm using a fine tuned mistral model.

Everytime I start the open ai server, with any request I get the 10 years old birthday prompt automatically injected.

It's very annoying, did anyone find a solution for this?

Tostino commented:
@Stealthwriter 
It got merged yesterday (the newer PR).

It'll be in the next release, or you can build/install from main now.
